### PR TITLE
Skip changing ids

### PR DESCRIPTION
**Pull request by janKozie1**

### CODE REVIEW

File: src/Highlight.ts [hunk 1]
```diff
@@ -32,11 +36,16 @@ export default class Highlight {
...
   public readonly range: Range;
+  public readonly options: IOptions;
   private data: IHighlightData;
   private _elements: HTMLElement[] = [];
 
-  constructor(range: Range, data: Pick<IHighlightData, Exclude<keyof IHighlightData, 'id'>> & Partial<Pick<IHighlightData, 'id'>>) {
+  constructor(
+    range: Range,
+    data: Pick<IHighlightData, Exclude<keyof IHighlightData, 'id'>> & Partial<Pick<IHighlightData, 'id'>>,
+    options?: IOptions) {
```
Comment (TomWoodward) [inline src/Highlight.ts]:
```suggestion
    options?: IOptions
  ) {
```

File: src/Highlight.ts [hunk 2]
```diff
@@ -54,7 +63,7 @@ export default class Highlight {
   }
 
   public removeStyle() {
-    const {style} = this.data;
+    const { style } = this.data;
```
Comment (TomWoodward) [inline src/Highlight.ts]: please avoid changing unrelated whitespace, it makes the diff harder to read, if it makes your life easier you can try using the `git add -p` option that allows you to stage chunks of diffs instead of whole files

File: src/Highlight.ts [hunk 3]
```diff
@@ -91,17 +100,48 @@ export default class Highlight {
   }
 
   public serialize(referenceElement?: HTMLElement): SerializedHighlight {
-    referenceElement = referenceElement || dom(this.range.commonAncestorContainer).closest('[id]');
+    const validReferenceElement = this.getValidReferenceElement(
+      this.options.skipIDsBy, referenceElement
```
Comment (TomWoodward) [inline src/Highlight.ts]: `getValidReferenceElement` is a class method that has access to `this.options`, so it seems unnecessary to pass it in

File: src/Highlight.ts [hunk 4]
```diff
@@ -91,17 +100,48 @@ export default class Highlight {
...
   }
   private loadStyle() {
-    const {style} = this.data;
+    const { style } = this.data;
     if (style) {
       this.elements.forEach((element) => element.classList.add(style));
     }
   }
+
+  private checkReferenceElement(validatingRegexps: RegExp[], referenceElement?: HTMLElement): boolean {
```
Comment (TomWoodward) [inline src/Highlight.ts]:
```suggestion
  private checkReferenceElement(validatingRegexps: RegExp[], referenceElement?: HTMLElement): referenceElement is HTMLElement {
```

will turn this into a guard so that when it passes ts will know its an html element and not undefined, and you can get rid of your `as HTMLElement` on line 139

File: src/Highlighter.ts [hunk 5]
```diff
@@ -11,6 +11,7 @@ interface IOptions {
   snapMathJax?: boolean;
   snapWords?: boolean;
   className?: string;
+  skipIDsBy?: RegExp[];
```
Comment (TomWoodward) [inline src/Highlighter.ts]: not sure its necessary to make this a list of regexes, we can just combine our criteria into one regex

File: src/Highlighter.ts [hunk 6]
```diff
@@ -145,7 +146,11 @@ export default class Highlighter {
         .filter((other: Highlight) => other.intersects(range));
 
       if (highlights.length === 0) {
-        const highlight: Highlight = new Highlight(range, {content: rangeContentsString(range)});
+        const highlight: Highlight = new Highlight(
+          range,
+          { content: rangeContentsString(range) },
+          { skipIDsBy: this.options.skipIDsBy }
```
Comment (TomWoodward) [inline src/Highlighter.ts]: can we create a new method `Highlighter.getHighlightOptions` which will return the subset of `options` that should be passed to the highlight constructor? after that can you also use it in `SerializedHighlight.load` when constructing highlights there? that way all highlights will have the same options regardless of if they were loaded from serialized data or made through the UI, and loading a highlight and serializing it again won't change the reference element

File: src/Highlight.ts (hunk 4 above)
Comment (janKozie1) [reply src/Highlight.ts]: But there is a chance that the `referenceElement` is a `HTMLElement` but it's not a valid one in which case i would need to access its parent. Am I missing something?

Comment (TomWoodward) [reply src/Highlight.ts]: my suggested change doesn't affect the conditions that you wrote for when the function returns true or false, it just says that when it _does_ return true, typescript will know that it is not undefined

File: src/Highlighter.ts (hunk 6 above)
Comment (TomWoodward) [reply src/Highlighter.ts]: can use getHighlightOptions here

File: src/Highlight.ts [hunk 7]
```diff
@@ -91,17 +101,48 @@ export default class Highlight {
...
   }
   private loadStyle() {
-    const {style} = this.data;
+    const { style } = this.data;
     if (style) {
       this.elements.forEach((element) => element.classList.add(style));
     }
   }
+
+  private checkReferenceElement(referenceElement?: HTMLElement): boolean {
```
Comment (TomWoodward) [inline src/Highlight.ts]: this conversation is not resolved https://github.com/openstax/highlighter/pull/31#discussion_r357793499

**Comment by janKozie1** [comment]
Sorry for requesting the review so late, I thought that reviewers get notifications when changes are committed
